Allow disabling per-day index (#6976)
VictoriaMetrics/VictoriaMetrics: docs/README.md

To override the default values see command-line flags with `-storage.cacheSize` prefix.
See the full description of flags [here](#list-of-command-line-flags).

## Index tuning

By default, VictoriaMetrics uses the following indexes for data retrieval:
`global` and `per-day`. Both store the same data, the only difference is that
the per-day index adds the date to each record. See [IndexDB](#indexdb) for
details.

If your use case involves
[high cardinality](https://docs.victoriametrics.com/faq/#what-is-high-cardinality)
with
[high churn rate](https://docs.victoriametrics.com/faq/#what-is-high-churn-rate)
then this default setting should be ideal for you.

A prominent example is Kubernetes. Its resources export a lot of metrics. The
resources are also created and deleted very often causing the metric label
values to change. This, in turn, results in new time series. The per-day index
speeds up data retrieval in this case.

But if your use case assumes low or no churn rate, then you might be benefiting
from disabling the per-day index by setting the flag `-disablePerDayIndex`. This
will improve the time series ingestion rate and decrease the disk space usage,

> since no time is spent to write to the per-day index and no disk space is used
> to store it.

Reviewer: ```suggestion
since no time or disk space is spent maintaining the per-day index.
```

Example use cases:

-   Historical weather data, such as
    [ERA5](https://cds.climate.copernicus.eu/cdsapp#!/dataset/reanalysis-era5-single-levels).
    It consists of millions time series whose hourly values span tens of years.

>     The time series	set never changes. If the per-day index is disabled, once

Reviewer: ```suggestion
    The time series set never changes. If the per-day index is disabled, once
```

-   IoT: a huge set of sensors exports time series with the sensor ID used as a
    metric label value. Since sensor additions or removals happen infrequently,
    the time series churn rate will be low. With the per-day index disabled, the
    entire time series set will be registered in global index during the initial
    data ingestion and the global index will receive small updates when a sensor
    is added or removed.

What to expect:

-   If the data was previously ingested with the per-day index enabled and then
    a search operation is performed with the per-day index disabled, the search
    result will still be correct, since both indexes (global and per-day)
    were populated at the ingestion time.

-   However, if the data was previously ingested with the per-day index disabled
    and then the search is performed with the per-day index enabled, the search
    result will be incorrect. It will either be empty or contain partial data
    (depending on whether data ingestion happened or not after the flag was set
    to false). This is because at the time of initial ingestion nothing was
    written	to the per-day index and at the search time the query is performed

>     on the per-day index.

Author: Done, please take a look.

since no time or disk space is spent maintaining the per-day index.

Example use cases:

-   Historical weather data, such as
    [ERA5](https://cds.climate.copernicus.eu/cdsapp#!/dataset/reanalysis-era5-single-levels).
    It consists of millions time series whose hourly values span tens of years.
    The time series set never changes. If the per-day index is disabled, once
    the first hour of data is ingested the entire time series set will be
    written into the global index and subsequent portions of data will not
    result in index update. But if the per-day index is enabled, the same set of
    time-series will be written to the per-day index for every day of data.

-   IoT: a huge set of sensors exports time series with the sensor ID used as a
    metric label value. Since sensor additions or removals happen infrequently,
    the time series churn rate will be low. With the per-day index disabled, the
    entire time series set will be registered in global index during the initial
    data ingestion and the global index will receive small updates when a sensor
    is added or removed.

What to expect:

-   Prefer setting this flag on fresh installations.

> -   Disabling per-day index with historical data is okay.

Reviewer: ```suggestion
-   Disabling per-day index on installations with historical data is Ok.
```

> -   Enabling per-day index with historical data will make it unsearcheable.

Reviewer: ```suggestion
-   Re-enabling per-day index on installations with historical data will make it unsearchable.
```

since no time or disk space is spent maintaining the per-day index.

Example use cases:

-   Historical weather data, such as
    [ERA5](https://cds.climate.copernicus.eu/cdsapp#!/dataset/reanalysis-era5-single-levels).
    It consists of millions time series whose hourly values span tens of years.
    The time series set never changes. If the per-day index is disabled, once
    the first hour of data is ingested the entire time series set will be
    written into the global index and subsequent portions of data will not
    result in index update. But if the per-day index is enabled, the same set of
    time-series will be written to the per-day index for every day of data.

-   IoT: a huge set of sensors exports time series with the sensor ID used as a
    metric label value. Since sensor additions or removals happen infrequently,
    the time series churn rate will be low. With the per-day index disabled, the
    entire time series set will be registered in global index during the initial
    data ingestion and the global index will receive small updates when a sensor
    is added or removed.

What to expect:

> 

Reviewer 2: Looks like this feature is recommended only for the fresh installations. If there were some data previously ingested, it will not be visible for search.

 Also, it's required to reset TSID cache if param was changed. Otherwise, it will not properly register index entries.

 Should we introduce some index search fallback to the previous index type based on requested time range?

 Or maybe it's better to add a check, if indexdb performs switch from 1 index type to another, directory must be empty?

Author: It is okay to use this feature for existing installations that previously had the per-day index enabled. In this case the search will use the global index must have all the data required for finding the data.

What is not okay is re-enabling per-day index after disabling it because in this case the the per-day index will be searched (for time ranges < 40 days) and the per day index will not have the data for the period when the per-day index has been disabled.

tsidCache does not store dates, so it shouldn't be reset. Please correct me if I am wrong.

I wouldn't introduce any checks or fallbacks. This feature is disabled by default and the documentation tells explictly what to expect. So the user should be knowing what they are doing in this case.

@hagen1778 / @valyala Do you have any comments on this?

Reviewer 2: > tsidCache does not store dates, so it shouldn't be reset. Please correct me if I am wrong.

 Yes, that's correct. Actually, it should be fine do not reset it. But in case if it will be reset later, it'll cause massive new index entries creation. And it'll increase size of the indexDB and require higher limits for max uniqueTimeSeries.

> I wouldn't introduce any checks or fallbacks. This feature is disabled by default and the documentation tells explictly what to expect. So the user should be knowing what they are doing in this case.

 Agree. Lets keep it as is.

Author: > But in case if it will be reset later, it'll cause massive new index entries creation.

Agree, this can happen. In this case, the ingestion will slow down until most of the metricNames are registered in global index.

> And it'll increase size of the indexDB

If there is no (or low) churn rate this will increase the index DB size by 1 day which will happen anyway every day if per-day index is enabled.

> ... and require higher limits for max uniqueTimeSeries.

Here too, if there is no churn rate, this limit shouldn't be affected.

Again, please let me know if I am missing something.

Reviewer 2: > Here too, if there is no churn rate, this limit shouldn't be affected.

 Not exactly. 

 For instance, you have zero churn rate and 1M Active Series (1M unique series). Before enabling this flag, query request may require up to 1M unique series limit. 
 After  disabling per-day index, an additional 1M of Active Series will be eventually registered at global index. And now it may return up to 2M unique series for search request. 

 In terms of disk usage, it'll be positive change. But from query prospective, it will have better performance only after 2nd index rotation.

Author: > After disabling per-day index, an additional 1M of Active Series will be eventually registered at global index. And now it may return up to 2M unique series for search request.

Agree. Thanks for finding this case. Just talked to @valyala, and he says it should be fine since not too many users will adopt that feature and out of those not too many will hit this limit.  Overall his recommendation is to not to address this case and update the documentation with the following:

1.  It is recommended to use this option with a completely new installation in order to avoid cases like this and
2.  Once the per-day index is disabled, it is not recommended to turn it on

Should we receive any complaints on exeeding the max unique ts, we can review the case and recommend to increase the value of maxUniqueTimeseries.

What do you think?

-   Disabling per-day index on data that has previously been ingested with
    enabled per-day index will on affect search results because the global index
    will be used for searching and it will contain all the necessary mappings to

>     data.

Reviewer: Will it have or not have effect? I can't get it from this description

Reviewer: Probably, we can simplify it to: "prefer setting this on fresh installation", "disabling it with historical would have bad impact", "enabling it with historical data is... okay?"

Author: Done. PTAL
